Sources/Foundation/JSONSerialization+Parser.swift

```diff
@@ -516,19 +516,42 @@ extension JSONParser {
         }
         
         // MARK: Numbers
-        
+
         private enum ControlCharacter {
             case operand
             case decimalPoint
             case exp
             case expOperator
         }
 
-        private mutating func parseNumber() throws -> String {
+
+        private func accumalateMagnitude(_ magnitude: UInt64?, value: UInt64) -> UInt64? {
+            guard let number = magnitude else { return nil }
+            let (newValue, overflow) = number.multipliedReportingOverflow(by: 10)
+            if overflow { return nil }
+            let (newMagnitude, overflow2) = newValue.addingReportingOverflow(value)
+            return overflow2 ? nil : newMagnitude
+        }
+
+        private func finalExponent(negativeExponent: Bool, exponentValue: Int?, fractionExponent: Int?) -> Int {
+            var result = (exponentValue ?? 0)
+            if negativeExponent { result = -1 * result }
+            result += (fractionExponent ?? 0)
+            return result
+        }
+
+
+        private mutating func parseNumber() throws -> _JSONNumber {
             var pastControlChar: ControlCharacter = .operand
             var numbersSinceControlChar: UInt = 0
             var hasLeadingZero = false
 
+            // Tracking digits for integer/exponent, eg parsing 12.34, exponent = -2 _integerMagnitude = 1234
+            var integerMagnitude: UInt64? = 0
+            var fractionExponent: Int?
+            var exponentValue: Int?
+            var negativeExponent = false
+            var jsonString = ""
```

Comment:
Thanks so much for moving your implementation into this `parseNumber` function! 👍 I guess the only thing left from my side here, is the creation of `jsonString`. In my point of view, we should only create the `jsonString` at the very end of the number parsing based on the numbers arraySlice.

```swift
self.makeStringFast(self[numberStartIndex ..< self.readerIndex])
```

Appending to a String especially with `Character(UnicodeScalar(ascii))` is, if I recall correctly, very slow.